File: ios-app/Doompedia/Data/WikipediaAPIClient.swift
import Foundation

struct OnlineArticle {
    let pageID: Int64
    let lang: String
    let title: String
    let summary: String
    let wikiURL: String
    let sourceRevID: Int64?
    let updatedAt: String
}

final class WikipediaAPIClient {
    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    func fetchRandomSummaries(language: String, count: Int) async throws -> [OnlineArticle] {
        let target = max(1, min(count, 40))
        var articles: [OnlineArticle] = []
        var seenTitles = Set<String>()
        let seed = Int(Date().timeIntervalSince1970 * 1000)

        for index in 0 ..< (target * 4) {
            if articles.count >= target { break }
            guard let url = URL(string: "https://\(language).wikipedia.org/api/rest_v1/page/random/summary?cb=\(seed + index)") else {
                break
            }
            let request = makeRequest(url: url)
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, (200 ... 299).contains(http.statusCode) else {
                continue
            }

            guard
                let payload = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                let title = payload["title"] as? String,
                let extract = payload["extract"] as? String,
                let page = payload["content_urls"] as? [String: Any],
                let desktop = page["desktop"] as? [String: Any],
                let pageURL = desktop["page"] as? String
            else {
                continue
            }

            if !seenTitles.insert(title).inserted { continue }

            let pageID = (payload["pageid"] as? NSNumber)?.int64Value ?? stablePageID(from: title)
            let timestamp = (payload["timestamp"] as? String) ?? ISO8601DateFormatter().string(from: Date())
            let revID = (payload["revision"] as? NSNumber)?.int64Value
            articles.append(
                OnlineArticle(
                    pageID: pageID,
                    lang: language,
                    title: title,
                    summary: extract,
                    wikiURL: pageURL,
                    sourceRevID: revID,
                    updatedAt: timestamp
                )
            )
        }

        return articles
    }

    func searchTitles(language: String, query: String, limit: Int) async throws -> [OnlineArticle] {
        let cleaned = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleaned.isEmpty else { return [] }
        let safeLimit = max(1, min(limit, 50))

        var components = URLComponents(string: "https://\(language).wikipedia.org/w/api.php")!
        components.queryItems = [
            URLQueryItem(name: "action", value: "opensearch"),
            URLQueryItem(name: "search", value: cleaned),
            URLQueryItem(name: "limit", value: "\(safeLimit)"),
            URLQueryItem(name: "namespace", value: "0"),
            URLQueryItem(name: "format", value: "json"),
            URLQueryItem(name: "formatversion", value: "2"),
        ]
        guard let url = components.url else { return [] }

        let request = makeRequest(url: url)
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200 ... 299).contains(http.statusCode) else {
            return []
        }

        guard let payload = try JSONSerialization.jsonObject(with: data) as? [Any],
              payload.count >= 4,
              let titles = payload[1] as? [String],
              let summaries = payload[2] as? [String],
              let urls = payload[3] as? [String]
        else {
            return []
        }

        let nowISO = ISO8601DateFormatter().string(from: Date())
        var output: [OnlineArticle] = []
        for index in titles.indices {
            let title = titles[index]
            guard index < urls.count else { continue }
            output.append(
                OnlineArticle(
                    pageID: stablePageID(from: title),
                    lang: language,
                    title: title,
                    summary: index < summaries.count ? summaries[index] : "",
                    wikiURL: urls[index],
                    sourceRevID: nil,
                    updatedAt: nowISO
                )
            )
        }
        return output
    }

    func fetchThumbnailURL(language: String, title: String, maxWidth: Int = 640) async throws -> String? {
        let cleaned = title.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleaned.isEmpty else { return nil }

        var components = URLComponents(string: "https://\(language).wikipedia.org/w/api.php")!
        components.queryItems = [
            URLQueryItem(name: "action", value: "query"),
            URLQueryItem(name: "format", value: "json"),
            URLQueryItem(name: "prop", value: "pageimages"),
            URLQueryItem(name: "piprop", value: "thumbnail"),
            URLQueryItem(name: "pithumbsize", value: "\(max(160, min(maxWidth, 1280)))"),
            URLQueryItem(name: "pilicense", value: "any"),
            URLQueryItem(name: "titles", value: cleaned),
        ]
        guard let url = components.url else { return nil }

        let request = makeRequest(url: url)
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200 ... 299).contains(http.statusCode) else {
            return nil
        }

        guard
            let payload = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let query = payload["query"] as? [String: Any],
            let pages = query["pages"] as? [String: Any]
        else {
            return nil
        }

        for (_, value) in pages {
            guard let page = value as? [String: Any],
                  let thumbnail = page["thumbnail"] as? [String: Any],
                  let source = thumbnail["source"] as? String,
                  !source.isEmpty
            else {
                continue
            }
            return source
        }
        return nil
    }

    private func makeRequest(url: URL) -> URLRequest {
        var request = URLRequest(url: url)
        request.timeoutInterval = 20
        request.setValue("Doompedia/0.1 (+https://github.com/mahmoudelfeelig/doompedia)", forHTTPHeaderField: "User-Agent")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        return request
    }

    private func stablePageID(from title: String) -> Int64 {
        var hash: UInt64 = 1469598103934665603
        for byte in title.utf8 {
            hash ^= UInt64(byte)
            hash &*= 1099511628211
        }
        return Int64(bitPattern: hash & 0x7fff_ffff_ffff_ffff)
    }
}

File: android-app/app/src/main/java/com/feelbachelor/doompedia/data/net/WikipediaApiClient.kt
package com.feelbachelor.doompedia.data.net

import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

data class OnlineArticle(
    val pageId: Long,
    val lang: String,
    val title: String,
    val summary: String,
    val wikiUrl: String,
    val sourceRevId: Long? = null,
    val updatedAt: String = "1970-01-01T00:00:00Z",
)

class WikipediaApiClient {
    fun fetchRandomSummaries(language: String, count: Int): List<OnlineArticle> {
        val normalizedLang = language.ifBlank { "en" }
        val cappedCount = count.coerceIn(1, 60)
        val seen = mutableSetOf<Long>()
        val results = mutableListOf<OnlineArticle>()
        val nowIso = java.time.Instant.now().toString()

        repeat(cappedCount * 2) {
            if (results.size >= cappedCount) return@repeat
            val endpoint = "https://$normalizedLang.wikipedia.org/api/rest_v1/page/random/summary"
            val payload = fetchJson(endpoint)
            val parsed = parseRandomSummary(payload, normalizedLang, nowIso) ?: return@repeat
            if (!seen.add(parsed.pageId)) return@repeat
            results += parsed
        }
        return results
    }

    fun searchTitles(language: String, query: String, limit: Int): List<OnlineArticle> {
        val normalizedLang = language.ifBlank { "en" }
        val cleanedQuery = query.trim()
        if (cleanedQuery.isBlank()) return emptyList()

        val cappedLimit = limit.coerceIn(1, 50)
        val nowIso = java.time.Instant.now().toString()
        val encodedQuery = URLEncoder.encode(cleanedQuery, StandardCharsets.UTF_8.name())
        val endpoint = "https://$normalizedLang.wikipedia.org/w/api.php" +
            "?action=opensearch&search=$encodedQuery&limit=$cappedLimit&namespace=0&format=json"

        val json = JSONArray(fetchJson(endpoint))
        val titles = json.optJSONArray(1) ?: JSONArray()
        val descriptions = json.optJSONArray(2) ?: JSONArray()
        val urls = json.optJSONArray(3) ?: JSONArray()

        val results = mutableListOf<OnlineArticle>()
        for (index in 0 until titles.length()) {
            val title = titles.optString(index).trim()
            if (title.isBlank()) continue
            val summary = descriptions.optString(index).trim().ifBlank { "$title article on Wikipedia" }
            val wikiUrl = urls.optString(index).trim().ifBlank {
                "https://$normalizedLang.wikipedia.org/wiki/" + URLEncoder.encode(
                    title.replace(' ', '_'),
                    StandardCharsets.UTF_8.name(),
                )
            }
            val pageId = stableSyntheticPageId(language = normalizedLang, title = title)
            results += OnlineArticle(
                pageId = pageId,
                lang = normalizedLang,
                title = title,
                summary = summary,
                wikiUrl = wikiUrl,
                sourceRevId = null,
                updatedAt = nowIso,
            )
        }
        return results
    }

    private fun parseRandomSummary(payload: String, language: String, nowIso: String): OnlineArticle? {
        val json = JSONObject(payload)
        val title = json.optString("title").trim()
        val summary = json.optString("extract").trim()
        if (title.isBlank() || summary.isBlank()) return null

        val pageId = json.optLong("pageid", -1L).takeIf { it > 0L }
            ?: stableSyntheticPageId(language = language, title = title)

        val urls = json.optJSONObject("content_urls")
        val wikiUrl = urls?.optJSONObject("desktop")?.optString("page").orEmpty()
            .ifBlank { urls?.optJSONObject("mobile")?.optString("page").orEmpty() }
            .ifBlank {
            json.optString("canonicalurl").ifBlank {
                "https://$language.wikipedia.org/wiki/" + URLEncoder.encode(
                    title.replace(' ', '_'),
                    StandardCharsets.UTF_8.name(),
                )
            }
        }

        return OnlineArticle(
            pageId = pageId,
            lang = language,
            title = title,
            summary = summary,
            wikiUrl = wikiUrl,
            sourceRevId = null,
            updatedAt = nowIso,
        )
    }

    private fun fetchJson(url: String): String {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 10_000
            readTimeout = 12_000
            setRequestProperty("Accept", "application/json")
            setRequestProperty("User-Agent", "Doompedia/0.1 (+https://github.com/mahmoudelfeelig/doompedia)")
        }
        return connection.use { conn ->
            val statusCode = conn.responseCode
            val body = if (statusCode in 200..299) {
                conn.inputStream
            } else {
                conn.errorStream ?: conn.inputStream
            }
            val text = body.bufferedReader().use(BufferedReader::readText)
            require(statusCode in 200..299) {
                "Wikipedia API error $statusCode"
            }
            text
        }
    }

    private fun stableSyntheticPageId(language: String, title: String): Long {
        val key = "$language:${title.lowercase()}"
        val hash = key.hashCode().toLong().and(0x7FFF_FFFFL)
        return 9_000_000_000L + hash
    }
}

private inline fun <T> HttpURLConnection.use(block: (HttpURLConnection) -> T): T {
    return try {
        block(this)
    } finally {
        disconnect()
    }
}
